File: npk-sensor/sensor-logic.cpp
#include <chrono>
#include <cstring>
#include <fcntl.h>
#include <memory>
#include <string>
#include <termios.h>
#include <thread>
#include <unistd.h>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "npk_sensor/srv/get_soil_data.hpp"

using GetSoilData = npk_sensor::srv::GetSoilData;

// ROS2 node that talks to an RS485/Modbus NPK soil sensor over UART
// and exposes the readings (pH, moisture, temp, EC, N/P/K) as a service.
class SoilSensorNode : public rclcpp::Node {
public:
    SoilSensorNode() : Node("soil_sensor_node") {
        // Open the serial connection to the sensor on startup
        serial_fd_ = open_serial("/dev/ttyAMA0");

        if (serial_fd_ < 0) {
            RCLCPP_ERROR(this->get_logger(), "Failed to open serial port /dev/ttyAMA0");
        } else {
            RCLCPP_INFO(this->get_logger(), "Opened serial port /dev/ttyAMA0");
        }

        // Advertise the "get_soil_data" service, handled by handle_request()
        service_ = this->create_service<GetSoilData>(
            "get_soil_data",
            std::bind(
                &SoilSensorNode::handle_request,
                this,
                std::placeholders::_1,
                std::placeholders::_2
            )
        );

        RCLCPP_INFO(this->get_logger(), "Soil Sensor Service Ready");
    }

    ~SoilSensorNode() {
        // Make sure we don't leak the file descriptor on shutdown
        if (serial_fd_ >= 0) {
            close(serial_fd_);
        }
    }

private:
    int serial_fd_;
    rclcpp::Service<GetSoilData>::SharedPtr service_;

    // Opens and configures the serial port for 9600 baud, 8N1, no flow control.
    // Returns the fd, or -1 on failure.
    int open_serial(const std::string &port) {
        int fd = open(port.c_str(), O_RDWR | O_NOCTTY | O_SYNC);

        if (fd < 0) {
            return -1;
        }

        struct termios tty;
        memset(&tty, 0, sizeof tty);

        if (tcgetattr(fd, &tty) != 0) {
            close(fd);
            return -1;
        }

        // Baud rate
        cfsetospeed(&tty, B9600);
        cfsetispeed(&tty, B9600);

        // 8 data bits, no parity, 1 stop bit (8N1), raw mode
        tty.c_cflag = (tty.c_cflag & ~CSIZE) | CS8;
        tty.c_iflag &= ~IGNBRK;
        tty.c_lflag = 0;   // no canonical mode, no echo, no signals
        tty.c_oflag = 0;   // no output processing

        // Non-blocking-ish read: return as soon as data's available,
        // but give up after 1 second (VTIME is in tenths of a second)
        tty.c_cc[VMIN] = 0;
        tty.c_cc[VTIME] = 10;

        tty.c_iflag &= ~(IXON | IXOFF | IXANY);   // no software flow control
        tty.c_cflag |= (CLOCAL | CREAD);          // enable receiver, ignore modem lines
        tty.c_cflag &= ~(PARENB | PARODD);        // no parity
        tty.c_cflag &= ~CSTOPB;                   // 1 stop bit
        tty.c_cflag &= ~CRTSCTS;                  // no hardware flow control

        if (tcsetattr(fd, TCSANOW, &tty) != 0) {
            close(fd);
            return -1;
        }

        return fd;
    }

    // Converts a space-separated hex string (e.g. "01 03 00 06") into raw bytes
    std::vector<uint8_t> hex_to_bytes(const std::string &hex) {
        std::vector<uint8_t> bytes;
        std::string clean;

        for (char c : hex) {
            if (c != ' ') {
                clean += c;
            }
        }

        for (size_t i = 0; i < clean.length(); i += 2) {
            std::string byte_string = clean.substr(i, 2);
            uint8_t byte = static_cast<uint8_t>(std::stoi(byte_string, nullptr, 16));
            bytes.push_back(byte);
        }

        return bytes;
    }

    // Sends a Modbus command (as a hex string) over serial and reads back
    // the response. expected_len is how many bytes we expect for this request.
    std::vector<uint8_t> send_request(const std::string &cmd_hex, int expected_len = 11) {
        std::vector<uint8_t> response;

        if (serial_fd_ < 0) {
            RCLCPP_ERROR(this->get_logger(), "Serial port is not open");
            return response;
        }

        std::vector<uint8_t> cmd = hex_to_bytes(cmd_hex);

        ssize_t written = write(serial_fd_, cmd.data(), cmd.size());

        if (written < 0) {
            RCLCPP_ERROR(this->get_logger(), "Serial write failed");
            return response;
        }

        // Give the sensor time to process and respond before we try to read
        std::this_thread::sleep_for(std::chrono::milliseconds(100));

        response.resize(expected_len);
        ssize_t bytes_read = read(serial_fd_, response.data(), expected_len);

        if (bytes_read <= 0) {
            RCLCPP_ERROR(this->get_logger(), "No response from NPK sensor");
            response.clear();
            return response;
        }

        // Trim the buffer down to however many bytes we actually got
        response.resize(bytes_read);
        return response;
    }

    // Extracts a single 16-bit value from a Modbus response (bytes 3-4 are
    // the data field after the address/function/byte-count header) and
    // scales it down by `div` (e.g. sensors that report pH*100 use div=100)
    float parse_value(const std::vector<uint8_t> &data, float div = 1.0f) {
        if (data.size() >= 7) {
            uint16_t val = (static_cast<uint16_t>(data[3]) << 8) | data[4];
            return static_cast<float>(val) / div;
        }

        return -1.0f;   // sentinel value meaning "no valid reading"
    }

    // Service callback: polls the sensor for every measurement one at a
    // time (separate Modbus register reads) and fills in the response.
    void handle_request(
        const std::shared_ptr<GetSoilData::Request> request,
        std::shared_ptr<GetSoilData::Response> response
    ) {
        (void)request;   // unused, request has no fields

        // Each of these is a pre-built Modbus RTU read-holding-register
        // command (address, function code, register, count, CRC) for a
        // specific sensor register.

        // 1. pH (register 0x0006, scaled by 100)
        auto res_ph = send_request("01 03 00 06 00 01 64 0B");
        float ph = parse_value(res_ph, 100.0f);

        // 2. Humidity / moisture (register 0x0012, scaled by 10)
        auto res_h = send_request("01 03 00 12 00 01 24 0F");
        float hum = parse_value(res_h, 10.0f);

        // 3. Temperature (register 0x0013, scaled by 10)
        auto res_t = send_request("01 03 00 13 00 01 75 CF");
        float temp = parse_value(res_t, 10.0f);

        // 4. Conductivity / EC (register 0x0015, no scaling)
        auto res_ec = send_request("01 03 00 15 00 01 95 CE");
        float ec = parse_value(res_ec, 1.0f);

        // 5. NPK - one request reads 3 consecutive registers (N, P, K)
        // starting at 0x001E, so we parse all three out of one response.
        auto res_npk = send_request("01 03 00 1E 00 03 65 CD");

        float n = -1.0f;
        float p = -1.0f;
        float k = -1.0f;

        if (res_npk.size() >= 11) {
            n = static_cast<float>((static_cast<uint16_t>(res_npk[3]) << 8) | res_npk[4]);
            p = static_cast<float>((static_cast<uint16_t>(res_npk[5]) << 8) | res_npk[6]);
            k = static_cast<float>((static_cast<uint16_t>(res_npk[7]) << 8) | res_npk[8]);
        }

        // Pack everything into the service response
        response->ph = ph;
        response->moisture = hum;
        response->temperature = temp;
        response->conductivity = ec;
        response->nitrogen = n;
        response->phosphorus = p;
        response->potassium = k;

        RCLCPP_INFO(this->get_logger(), "Soil data returned");
    }
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SoilSensorNode>();
    rclcpp::spin(node);   // block here, servicing requests until shutdown
    rclcpp::shutdown();
    return 0;
}

File: npk-sensor/sensor-logic.py
import rclpy
from rclpy.node import Node
import serial
import time

from npk_sensor.srv import GetSoilData


# ROS2 node that communicates with the NPK sensor over RS485/Modbus
# and provides the sensor readings through a ROS service
class SoilSensorNode(Node):

    def __init__(self):
        super().__init__('soil_sensor_node')

        # Open the serial connection to the sensor
        self.ser = serial.Serial(
            '/dev/ttyAMA0',
            9600,
            timeout=1
        )

        # Create the service that other ROS nodes can call
        # to get the current soil sensor readings
        self.srv = self.create_service(
            GetSoilData,
            'get_soil_data',
            self.handle_request
        )

        self.get_logger().info("Soil Sensor Service Ready")

    # Send a Modbus command to the sensor and return its response
    def send_request(self, cmd_hex):
        try:
            self.ser.write(bytes.fromhex(cmd_hex))

            # Give the sensor time to respond
            time.sleep(0.1)

            return self.ser.read(11)

        except Exception as e:
            self.get_logger().error(f"Comm error: {e}")
            return None

    # Read a single value from the sensor response
    # The divisor is used for values that the sensor reports with scaling
    def parse_value(self, data, div=1):
        if data and len(data) >= 7:
            val = int.from_bytes(
                data[3:5],
                byteorder='big'
            )

            return val / div

        return None

    # This runs whenever another ROS node calls the soil data service
    def handle_request(self, request, response):

        # Read pH
        res_ph = self.send_request(
            "01 03 00 06 00 01 64 0B"
        )
        ph = self.parse_value(res_ph, 100)

        # Read humidity
        res_h = self.send_request(
            "01 03 00 12 00 01 24 0F"
        )
        hum = self.parse_value(res_h, 10)

        # Read temperature
        res_t = self.send_request(
            "01 03 00 13 00 01 75 CF"
        )
        temp = self.parse_value(res_t, 10)

        # Read conductivity
        res_ec = self.send_request(
            "01 03 00 15 00 01 95 CE"
        )
        ec = self.parse_value(res_ec, 1)

        # Read nitrogen, phosphorus, and potassium
        # All three values are returned in the same response
        res_npk = self.send_request(
            "01 03 00 1E 00 03 65 CD"
        )

        n = p = k = None

        if res_npk and len(res_npk) >= 11:
            n = int.from_bytes(
                res_npk[3:5],
                byteorder='big'
            )

            p = int.from_bytes(
                res_npk[5:7],
                byteorder='big'
            )

            k = int.from_bytes(
                res_npk[7:9],
                byteorder='big'
            )

        # Put the sensor readings into the ROS response
        # -1.0 is returned if a reading could not be obtained
        response.ph = float(ph) if ph is not None else -1.0
        response.moisture = float(hum) if hum is not None else -1.0
        response.temperature = float(temp) if temp is not None else -1.0
        response.conductivity = float(ec) if ec is not None else -1.0
        response.nitrogen = float(n) if n is not None else -1.0
        response.phosphorus = float(p) if p is not None else -1.0
        response.potassium = float(k) if k is not None else -1.0

        self.get_logger().info("Soil data returned")

        return response

    # Close the serial connection when the node shuts down
    def destroy_node(self):
        self.ser.close()
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)

    node = SoilSensorNode()

    # Keep the node running and wait for service requests
    rclpy.spin(node)

    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
